### ISSUE TITLE
If git_icons = true, really slow.

### ISSUE THREAD

**Issue by uga-rosa** [ibhagwan/fzf-lua#126]
This setting alone will slow down the speed to open `FzfLua files` by more than a second.
Do you have any idea why?

NVIM v0.5.0 (stable)
OS: Ubuntu on WSL2
git version 2.25.1

**Comment by ibhagwan** [comment]
In order to determine the icons for files a couple of commands are run in the background, it’s possible one of them is slow, can you run the below and let me know if they take a significant time?

```lua
    git_diff_cmd        = "git diff --name-status --relative HEAD",
    git_untracked_cmd   = "git ls-files --exclude-standard --others",
```

**Comment by uga-rosa** [comment]
When run in a terminal, these commands returned values fast...

**Comment by ibhagwan** [comment]
> When run in a terminal, these commands returned values fast...

Can you try running `files` with `git_icons = true` in a new repo with just a couple of files or on different repos and see if this happens on all git repos or just specific ones? I’m trying to think what else could be slowing things down other than the commands above.

**Comment by uga-rosa** [comment]
There are three levels that I can feel. The fastest is when I set git_icons = false; even outside the git repo, setting git_icons = true slows it down a bit. The slowest was when I was inside the git repo and set git_icons = true. This happens even in a newly created repo with no files in it.

**Comment by ibhagwan** [comment]
Perhaps the command that checks if you’re in a git repo slows it down, can you run this in the shell in the same locations you ran fzf-lua and let me know if this seems like the culprit?

```sh
git rev-parse --show-toplevel
```

**Comment by ibhagwan** [comment]
There are the only 3 extra commands running when you specify `git_icons`, maybe it has something to do with being run inside nvim?

Can you try running the commands inside neovim like this:

```lua
:lua print(vim.fn.system('<command>'))
```

**Comment by uga-rosa** [comment]
I feel a slight delay, but I am not sure if it is due to `print`.
I've also looked at the source, and I can't think of any reason other than the slow call to vim.fn.systemlist :(

**Comment by ibhagwan** [comment]
> I feel a slight delay, but I am not sure if it is due to `print`.
> I've also looked at the source, and I can't think of any reason other than the slow call to vim.fn.systemlist :(

Yes, this is awfully strange, there really isn’t much code added for the git icons.

if you suspect `systemlist` how about we try with either the [official neovim 0.5 appimage](https://github.com/neovim/neovim/releases/download/v0.5.0/nvim.appimage) or the nightly?

**Comment by uga-rosa** [comment]
I use it (v0.5.0). I will try nightly once.

**Comment by uga-rosa** [comment]
There was no improvement.
Can it be improved by running external commands directly from lua? I'll fork it later and give it a try.

**Comment by ibhagwan** [comment]
Let me know how it goes or if you could find the source of the delay in your system.

**Comment by jdrouhard** [comment]
Yes, `git_icons = true` is slow in larger git repositories. I experience the same problem, but unfortunately I have not been able to come up with a good solution for it yet. If you don't care about untracked files specifically, you can set the `git_untracked_cmd` to `'false'` in your config, which helps some. Trade-off is you lose the `?` for untracked files, but it should speed it up a bit. Still won't be as fast as disabling git icons altogether.

```lua
files = {
    git_untracked_cmd = 'false',
}
```

We can't run the git commands in parallel because we have to have the results before the filter command is called for each entry (which is done prior to piping to fzf).

**Comment by jdrouhard** [comment]
> Can it be improved by running external commands directly from lua? I'll fork it later and give it a try.

Please let us know how this goes! It could definitely be `vim.fn.systemlist` that's causing a slowdown here. You might be able to use `vim.loop` directly to spawn the processes and read the output, though it would be quite a bit more verbose. Worth it if it's faster, though :)

**Comment by ibhagwan** [comment]
> Yes, git_icons = true is slow in larger git repositories

@uga-rosa is also saying it’s slower in his case in a clean repository or even outside of a git repo:

>There are three levels that I can feel. The fastest is when I set git_icons = false; even outside the git repo, setting git_icons = true slows it down a bit. The slowest was when I was inside the git repo and set git_icons = true. This happens even in a newly created repo with no files in it.

I’m not sure how can we make the git repo detection command faster, I wonder if there would be a difference between the two commands below in terms of speed:

```sh
git rev-parse --show-toplevel         # what is currently being used
git rev-parse --is-inside-work-tree   # will this a tiny bit faster?
```

**Comment by jdrouhard** [comment]
I've spent a lot of time trying to get this to be as fast as possible with a git repo I work with daily, though most of my time was spent optimizing the lua code that processes the entry and generates the result. I don't think choosing different git commands will make much of a difference, here. I think the problem are the three commands it runs when the icons are enabled (or the fact it has to run commands at all). There might be quite a bit of overhead to `vim.fn.systemlist()` or `vim.fn.system()` like @uga-rosa suggested.

Slight delays in each of these compound and make it feel perceptibly slower. Each millisecond counts unfortunately, and I've had to disable icons for a few providers myself.

**Comment by ibhagwan** [comment]
> I've had to disable icons for a few providers myself.

The eye candies come at a price I suppose.

**Comment by uga-rosa** [comment]
vim.loop.spawn runs asynchronously, so io.popen would be more appropriate for this application. I'm still at work, so I haven't tried it yet, but...

**Comment by ibhagwan** [comment]
> vim.loop.spawn runs asynchronously, so io.popen would be more appropriate for this application. I'm still at work, so I haven't tried it yet, but...

But then if the result returns after the find command processed it’s entry  it won’t be able to update the entries inside fzf.

Although I could do some complex gymnastics to make it work I wouldn’t want to add such complexity solely for the git icons delay, you can always choose to run it without icons and when you wish to see the git icons run a specialized command with `git_icons = true` or use the `git_status` provider.

**Comment by uga-rosa** [comment]
> But then if the result returns after the find command processed it’s entry it won’t be able to update the entries inside fzf.

I understand. I talked that vim.loop might cause it, that it would be better not to use.
Sorry for my bad English.

**Comment by jdrouhard** [comment]
I benchmarked using `io.popen` instead of `systemlist` and there was no difference at all, unfortunately. The only way to maybe squeeze a bit of performance out of this is to start an async job that runs the git commands, and then wait for the job to finish the first time the results are needed in `make_entry_file()`. Not sure it would help much or be worth the complexity but I may play with that later anyway.

**Comment by uga-rosa** [comment]
In my case, this change clearly made it faster. However, it is not enough.

fzf-lua/stdio.lua
```lua
local M = {}

M.get_stdout = function(cmd)
  local stdout = {}
  for h in io.popen(cmd, "r"):lines() do
    stdout[#stdout + 1] = h
  end
  return stdout
end

return M
```

fzf-lua/core.lua
```lua
local stdio = require "fzf-lua.stdio"

~

local get_diff_files = function(opts)
    local diff_files = {}

    local status = stdio.get_stdout(path.git_cwd(
      config.globals.files.git_diff_cmd, opts.cwd))
    for _, v in ipairs(status) do
      local icon, file = v:match("^([MUDAR])%s+(.*)")
      if icon and file then
        diff_files[file] = icon
      end
    end

    --local status = vim.fn.systemlist(path.git_cwd(
    --  config.globals.files.git_diff_cmd, opts.cwd))
    --if not utils.shell_error() then
    --    for i = 1, #status do
    --      local icon, file = status[i]:match("^([MUDAR])%s+(.*)")
    --      if icon and file then diff_files[file] = icon end
    --    end
    --end

    return diff_files
end

local get_untracked_files = function(opts)
    local untracked_files = {}

    local status = stdio.get_stdout(path.git_cwd(
      config.globals.files.git_untracked_cmd, opts.cwd))

    for _, v in ipairs(status) do
      untracked_files[v] = "?"
    end

    --local status = vim.fn.systemlist(path.git_cwd(
    --  config.globals.files.git_untracked_cmd, opts.cwd))
    --if vim.v.shell_error == 0 then
    --    for i = 1, #status do
    --        local file = status[i]
    --        untracked_files[file] = "?"
    --    end
    --end

    return untracked_files
end
```

**Comment by uga-rosa** [comment]
By the way, this is a suggestion, why not use stylua as the formatter?
By using stylua.toml, we can use a common setting for each project.

**Comment by ibhagwan** [comment]
> In my case, this change clearly made it faster. However, it is not enough.

How much faster, was it noticeable? Are you still looking into a 1+ second delay? My machine is from 2015 and it’s pretty instantaneous for me. What are your hardware specs?

> By the way, this is a suggestion, why not use stylua as the formatter?
By using stylua.toml, we can use a common setting for each project.

I’m pretty old school, never liked the idea of something that automatically modifies my code (prettier, etc), I might be “missing out” I just never opted for usage of such tools.

**Comment by uga-rosa** [comment]
It was less than a second, maybe 0.7 seconds (in fzf-lua repo).
I think hardware specs are not low. i5-9400, RAM 16GB, M.2 SSD.

> I’m pretty old school, never liked the idea of something that automatically modifies my code (prettier, etc), I might be “missing out” I just never opted for usage of such tools.

I see. I will respect your style.

**Comment by ibhagwan** [comment]
> I see. I will respect your style.

Whatever you do is fine, I mostly used `indent=2` except for leftover original lines of code before I reduced from `4->2`, whatever your style is I will happily accept it.

**Comment by ibhagwan** [comment]
> I think hardware specs are not low. i5-9400, RAM 16GB, M.2 SSD.

That’s very reasonable, I’m running Linux, are you on MacOS? If we can isolate this issue to `vim.fn.system` calls on Mac that would be helpful.

**Comment by ibhagwan** [comment]
Last spam question, are you by any chance using [vim-fugitive](https://github.com/tpope/vim-fugitive)? If so, how fast does it take to open `:Git`?

I haven’t looked into it’s code but maybe it has some sort of caching or is somehow faster (with some magical tpope trick?) - if so I can try to auto-detect if you have fugitive installed and if so perhaps use it’s interface to query the icons.

**Comment by uga-rosa** [comment]
I don't use it.

**Comment by uga-rosa** [comment]
OS is Ubuntu on WSL2, not macOS.

**Comment by ibhagwan** [comment]
> I don't use vim-fugitive.

Might be worth a quick test? It’s been around forever and is made by one of the greatest vimmers of all time - if it works fast for you I’ll dive deeper into the code.

**Comment by jdrouhard** [comment]
> Might be worth a quick test? It’s been around forever and is made by one of the greatest vimmers of all time - if it works fast for you I’ll dive deeper into the code.

I do use vim-fugitive, but running `:Git` still has a noticeable delay before it comes up so I don't think there's any magic trickery there to make it work faster.

Our best bet is to play around with spawning a job that runs the commands immediately when `fzf_files()` is first called, and then waiting on the results in the first call to `make_entry_file()` which isn't called until nvim-fzf has had a chance to set up the pipes and run the provider command. That lets there be a bit of parallelization with git and nvim-fzf setting up the command and pipes.

**Comment by uga-rosa** [comment]
I was able to speed up the process :)
It may have been due to error handling.

**Comment by ibhagwan** [comment]
> I was able to speed up the process :)

Do share :)

**Comment by uga-rosa** [comment]
I'm creating PR. Please a wait.

**Comment by uga-rosa** [comment]
I also fixed some problems that I noticed. For example, `vim.fn.systemlist()` returns an empty string when it fails, so there is no point in displaying it as an error message. So, I made it so that it displays the command that was executed.

**Comment by uga-rosa** [comment]
With this change, I can now open it in an almost imperceptible speed.

**Comment by ibhagwan** [comment]
> With this change, I can now open it in an almost imperceptible speed.

Tysm! I never had any delay so I can’t measure it I’m wondering what is  @jdrouhard experiencing with this change?

**Comment by uga-rosa** [comment]
I took a gif to see how much it had changed.

![fzflua](https://user-images.githubusercontent.com/82267684/132959159-ad2d51ba-fe9a-452b-a22e-c9831aa44a9b.gif)

**Comment by ibhagwan** [comment]
> I took a gif to see how much it had changed.
> 
> ![fzflua](https://user-images.githubusercontent.com/82267684/132959159-ad2d51ba-fe9a-452b-a22e-c9831aa44a9b.gif)

I don’t know how slow it was for you before the change but this is looking very crisp!

**Comment by uga-rosa** [comment]
I used `vim.fn.systemlist` for the first one and `io.popen` for the second one. there is a delay after pressing enter, but it's hard to tell...

**Comment by ibhagwan** [comment]
https://github.com/ibhagwan/fzf-lua/commit/8871e7485a6127c1a5d47a1fe93991000cdfcd6c

Implemented this for you as an experimental feature, enable it by adding `lua_io = true` to the setup function (same as your closed PR).

If you set it right you should be seeing the below message during the setup of fzf-lua:
```sh
[Fzf-lua] using experimental feature 'lua_io'
```
No other changes required, let me know if this speeds things up for you?

Btw, I labeled this experimental for the time being due to the hacky nature of getting stderr/exit code from `io.popen`.

**Comment by uga-rosa** [comment]
Tysm! It's faster!

**Comment by uga-rosa** [comment]
Oh, I found the cause. This is ridiculous...
zsh is slow to start, I changed shell to bash and it's faster...

**Comment by ibhagwan** [comment]
> Oh, I found the cause. This is ridiculous...
> zsh is slow to start, I changed shell to bash and it's faster...

You’re right! When I tested the code I noticed `io.popen` started `sh` while `systemlist` was using `vim.env.SHELL`.

**Comment by uga-rosa** [comment]
Is it possible to select the shell to use with fzf-lua (like setlocal shell="/bin/bash")?

**Comment by ibhagwan** [comment]
> Is it possible to select the shell to use with fzf-lua (like setlocal shell="/bin/bash")?

I can probably do that for the fzf window but it won’t be of any help here, the git commands are all running before fzf is open, obviously I can change `vim.env.SHELL` but that would override the user’s neovim config which I would like to avoid.

**Comment by ibhagwan** [comment]
Anyways I think It’s safe to close this issue now, glad we were able to get to the bottom of it.

**Comment by ibhagwan** [comment]
Sorry to ressurrect this old thing but I just recently learned the following on neovim's `system` (and `systemlist`) commands:

```help
system({cmd} [, {input}])				*system()* *E677*
  Gets the output of {cmd} as a |string| (|systemlist()| returns
  a |List|) and sets |v:shell_error| to the error code.

  {cmd} is treated as in |jobstart()|:
  If {cmd} is a List it runs directly (no 'shell').
  If {cmd} is a String it runs in the 'shell', like this: >
    :call jobstart(split(&shell) + split(&shellcmdflag) + ['{cmd}'])
```

Which means that `vim.fn.system("cmd arg1 arg2")` is fundementally different than `vim.fn.system({"cmd", "arg1", "arg2"})`.

After commenting [on this reddit post](https://www.reddit.com/r/neovim/comments/r05u16/utf8_library_for_neovim_lua/hlqn9z6/?context=3), I remembered where I know @uga-rosa from and this issue just gave me the push that I needed to make these improvements.

https://github.com/ibhagwan/fzf-lua/commit/abd1a3cc90d7fa2680036f2b871de6088dbfc4ca

This commit makes sure that almost all commands that can be sent as tables are sent that way, most importantly this affects the `git status` and `git rev-parse` (used to determine git root and check if git repo) so no matter what shell you use in neovim it wouldn't affect the speed of the popup opening.
